runtime/jvmMain/src/kotlinx/serialization/internal/Platform.kt

```diff
@@ -53,12 +53,16 @@ internal actual fun <T : Any> KClass<T>.constructSerializerForGivenTypeArgs(vara
     return polymorphicSerializer()
 }
 
+@Suppress("UNCHECKED_CAST")
+private fun <A: Annotation, T: Any> Class<T>.getDeclaredAnnotation(annotationClass: KClass<A>) =
+    declaredAnnotations.firstOrNull { it.annotationClass == annotationClass } as? A
```

Comment:
I think there is no need to iterate through all the possible annotations, because it may be quite a slow process. Moreover, [`getDeclaredAnnotations()` method](https://docs.oracle.com/javase/8/docs/api/java/lang/reflect/AnnotatedElement.html#getDeclaredAnnotations--) is also marked as 'since Java 1.8', so it does not solve the original problem.

I think what should be done here is to replace `getDeclaredAnnotation` with regular [`getAnnotation`](https://docs.oracle.com/javase/8/docs/api/java/lang/reflect/AnnotatedElement.html#getAnnotation-java.lang.Class-). 

The only difference between them, according to https://docs.oracle.com/javase/8/docs/api/java/lang/reflect/AnnotatedElement.html , is that `getDeclaredAnnotation` ignores `@Inherited` annotations from superclasses. Since no annotation in kotlinx.serialization is inheritable, there should be no difference.

Reply:
The problem is that android does not support Java 8 fully.
The reason behind the change is that [`getDeclaredAnnotations`](https://developer.android.com/reference/java/lang/reflect/AnnotatedElement#getDeclaredAnnotations()) is onle available for Android 24 and above while [`getDeclaredAnnotation`](https://developer.android.com/reference/java/lang/reflect/AnnotatedElement#getDeclaredAnnotation(java.lang.Class%3CT%3E)`) is available on all android versions.

The performance shouldn't be an issue too, because if you take a look at the implementation of `getDeclaredAnnotation`, it internally calls `getDeclaredAnnotations` so there's no deifference.

[Android](https://android.googlesource.com/platform/libcore/+/refs/tags/android-7.0.0_r1/ojluni/src/main/java/java/lang/Class.java#2395)
```java
@Override public Annotation[] getAnnotations() {
    HashMap<Class<?>, Annotation> map = new HashMap<Class<?>, Annotation>();
    for (Annotation declaredAnnotation : getDeclaredAnnotations()) {
        map.put(declaredAnnotation.annotationType(), declaredAnnotation);
    }
    for (Class<?> sup = getSuperclass(); sup != null; sup = sup.getSuperclass()) {
        for (Annotation declaredAnnotation : sup.getDeclaredAnnotations()) {
            Class<? extends Annotation> clazz = declaredAnnotation.annotationType();
            if (!map.containsKey(clazz) && clazz.isDeclaredAnnotationPresent(Inherited.class)) {
                map.put(clazz, declaredAnnotation);
            }
        }
    }
    Collection<Annotation> coll = map.values();
    return coll.toArray(new Annotation[coll.size()]);
}

```

[OpenJDK 8](http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/687fd7c7986d/src/share/classes/java/lang/Class.java#l3344)
```java
public <A extends Annotation> A getDeclaredAnnotation(Class<A> annotationClass) {
    Objects.requireNonNull(annotationClass);

    return (A) annotationData().declaredAnnotations.get(annotationClass);
}
```

Reply:
The problem is that android does not support Java 8 fully.
The reason behind the change is that [`getDeclaredAnnotations`](https://developer.android.com/reference/java/lang/reflect/AnnotatedElement#getDeclaredAnnotations()) is onle available for Android 24 and above while [`getDeclaredAnnotation`](https://developer.android.com/reference/java/lang/reflect/AnnotatedElement#getDeclaredAnnotation(java.lang.Class%3CT%3E)`) is available on all android versions.

The performance shouldn't be an issue too, because if you take a look at the implementation of `getDeclaredAnnotation`, it internally calls `getDeclaredAnnotations` so there's no deifference.

[OpenJDK 8](http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/687fd7c7986d/src/share/classes/java/lang/Class.java#l3344)
```java
public <A extends Annotation> A getDeclaredAnnotation(Class<A> annotationClass) {
    Objects.requireNonNull(annotationClass);

    return (A) annotationData().declaredAnnotations.get(annotationClass);
}
```

Reply:
Indeed, I misread the description: `getDeclaredAnnotations()` is available in Java 1.5 and Android API 1. However, aforementioned `annotationData().declaredAnnotations` is a map, so while `getDeclaredAnnotations()` allocates an additional array and puts map contents to it, `getDeclaredAnnotation` simply calls `Map.get`. So I'd still rather stick with `getAnnotation` instead of `getDeclaredAnnotations`.

Reply:
Hmm, I don't think the array of `Annotation` will take up that much memory even if there was several hundred annotation in a class to be honest

Reply:
It's more about unnecessary allocation/deallocation and GC work

Reply:
oops, I thought you're talking about `getDeclaredAnnotation` 😅 
the names are confusing, aren't they?
Also didn't notice that `getAnnotation` existed since API level 1.

Fixed.